sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsWithStateSuite.scala

```diff
@@ -1268,12 +1269,298 @@ class FlatMapGroupsWithStateSuite extends StateStoreMetricsTest {
+      }
+    )
+  }
+
+  test("flatMapGroupsWithState - initial state - java api") {
+    val initialState = Seq(
+      ("a", 2)
+    ).toDS().groupByKey(_._1).mapValues(_._2)
+
+    val javaStateFunc = new FlatMapGroupsWithStateFunction[String, String, Int, String] {
+      import java.util.{Iterator => JIterator}
+      override def call(
+          key: String,
+          values: JIterator[String],
+          state: GroupState[Int]): JIterator[String] = {
+        state.update(0)
+        new JIterator[String] {
+          override def hasNext: Boolean = false
+          override def next(): String = null
+        }
+      }
+    }
+
+    val inputData = MemoryStream[String]
+    val result = inputData.toDS().groupByKey(x => x).flatMapGroupsWithState(
+      javaStateFunc, OutputMode.Update,
+      implicitly[Encoder[Int]], implicitly[Encoder[String]],
+      GroupStateTimeout.NoTimeout, initialState)
+    testStream(result, Update)(
+      AddData(inputData, "b"),
+      CheckNewAnswer(),
+      assertNumStateRows(total = 2, updated = 2)
+    )
+  }
+
+  test("mapGroupsWithState - initial state - java api") {
+    val initialState = Seq(
+      ("a", 4),
+      (null, 2)
+    ).toDS().groupByKey(_._1).mapValues(_._2)
+
+    val javaStateFunc = new MapGroupsWithStateFunction[String, String, Int, String] {
+      import java.util.{Iterator => JIterator}
+      override def call(
+          key: String,
+          values: JIterator[String],
+          state: GroupState[Int]): String = {
+        var valSize = 0
+        while (values.hasNext) {
+          valSize += 1
+          values.next()
+        }
+        val count = state.getOption.getOrElse(0)
+        state.update(count + valSize)
+        state.get.toString
+      }
+    }
+
+    val inputData = MemoryStream[String]
+    val result = inputData.toDS().groupByKey(x => x).mapGroupsWithState(
+      javaStateFunc,
+      implicitly[Encoder[Int]], implicitly[Encoder[String]],
+      GroupStateTimeout.NoTimeout, initialState)
+    testStream(result, Update)(
+      AddData(inputData, null),
+      CheckNewAnswer("4", "3"),
+      assertNumStateRows(total = 2, updated = 2)
+    )
+  }
+
+  test("flatMapGroupsWithState - initial state - streaming initial state") {
+    val initialStateData = MemoryStream[(String, RunningCount)]
+    initialStateData.addData(("a", new RunningCount(1)))
+
+    val inputData = MemoryStream[String]
+
+    val result =
+      inputData.toDS()
+        .groupByKey(x => x)
+        .flatMapGroupsWithState(
+          Update, NoTimeout(), initialStateData.toDS().groupByKey(_._1).mapValues(_._2)
+        )(flatMapGroupsWithStateFunc)
+
+    val e = intercept[AnalysisException] {
+      result.writeStream
+        .format("console")
+        .start()
+    }
+    assert(e.message.contains("Initial state cannot be a streaming DataFrame/Dataset."))
+  }
+
+  testWithAllStateVersions("mapGroupsWithState - initial state - test api") {
```

Comment:
what is the unit test for?

Reply:
just to have coverage on the mapGroupsWithState api I think I can condense it with the null key test